src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -94,17 +127,16 @@ public async Task<ProjectSystemProject> CreateAndAddToWorkspaceAsync(
         cancellationToken = CancellationToken.None;
 #pragma warning restore IDE0059 // Unnecessary assignment of a value
 
-        _visualStudioWorkspaceImpl.ProjectSystemProjectFactory.SolutionPath = solutionFilePath;
+        _visualStudioWorkspaceImpl.ProjectSystemProjectFactory.SolutionPath = _solutionFilePath;
         _visualStudioWorkspaceImpl.ProjectSystemProjectFactory.SolutionTelemetryId = GetSolutionSessionId();
 
-        var hostInfo = new ProjectSystemHostInfo(_dynamicFileInfoProviders, vsixAnalyzerProvider, _analyzerAssemblyRedirectors);
-        var project = await _visualStudioWorkspaceImpl.ProjectSystemProjectFactory.CreateAndAddToWorkspaceAsync(projectSystemName, language, creationInfo, hostInfo);
+        var hostInfo = new ProjectSystemHostInfo(_dynamicFileInfoProviders, _vsixAnalyzerProvider!, _analyzerAssemblyRedirectors);
+        var project = await _visualStudioWorkspaceImpl.ProjectSystemProjectFactory.CreateAndAddToWorkspaceAsync(projectSystemName, language, creationInfo, hostInfo).ConfigureAwait(true);
 
         _visualStudioWorkspaceImpl.AddProjectToInternalMaps(project, creationInfo.Hierarchy, creationInfo.ProjectGuid, projectSystemName);
 
         // Ensure that other VS contexts get accurate information that the UIContext for this language is now active.
-        // This is not cancellable as we have already mutated the solution.
-        await _visualStudioWorkspaceImpl.RefreshProjectExistsUIContextForLanguageAsync(language, CancellationToken.None);
+        _uiContextUpdateWorkQueue.AddWork(language);
```

Comment:
I'm not sure if asynchronously updating this will cause issues, or whether we had a (soft) guarantee that when we return the context would be set. But there might be an even better approach here: we only need to update the contexts when the number of projects will go from zero to non-zero during a load. Which in our case of loading Roslyn as a benchmark would be a grand total of two. So maybe a small tweak to simply have it realize the update isn't necessary in the first place might be even better.

Reply:
It feels like the only way to know whether the update is necessary is to check the state of the UIContext, which would need to be done on the UI thread. It doesn't seem like just keeping track if we go to/from zero projects with the language name would handle cases where other code could change that UIContext's state.

Reply:
I'm not seeing where this is even used (other than tests). Is there somewhere with IVT that is using these flags?

Reply:
@jasonmalinowski -- do you have a preference between:

1) Loosen the requirement that uicontext.IsActive is set during CreateAndAddToWorkspaceAsync execution. Change the integration tests that look like they might rely on this to instead do a WaitForAllAsyncOperationsAsync.

2) Change RefreshProjectExistsUIContextForLanguageAsync to assume that no one else changes the UIContext.IsActive state and have it set the value when the number of projects for a language switches to/from zero.

3) Something else

I personally prefer 1 since I can't find any other users of these UIContexts, but I'm not too opinionated.

Reply:
The UI contexts are used for things like enabling/disabling commands, and also to help with other components that want to be able to ask "is there a C# project even there before I load up all my Roslyn support?" for things like test explorer. We're the only one who should be setting it so this assumption:

>  to assume that no one else changes the UIContext.IsActive state

Is completely fine to assume I'd say.

I'd be worried about weakening the support because it could have downstream ramifications -- package loads and such can trigger on UIContexts being changed, and delaying that on some time delay could break integration tests.

Reply:
That said, if 2 is hard to write, maybe 1 is just the easier option.

Reply:
I missed this question:

> Is there somewhere with IVT that is using these flags?

Anything can use the GUID for the context to ask if that context is active. I'm not sure if the GUIDs got formally documented but any other repo could have it.

Reply:
Tried out 2), let me know if you were thinking along a different line for it. 

Note that only calls to CreateAndAddToWorkspaceAsync that will change the UIContext state wait for it to update before returning. Is this acceptable?